### ISSUE TITLE
Should be possible to get the "qualified" stack name

### ISSUE THREAD

**Issue by hausdorff** [pulumi/pulumi#2800]
`getStack`, as implemented in both the Python and Node.js SDKs return only the stack name itself. They do not return the organization name.

I'm not sure what we're using this for, but it seems to me that it would be _much_ more useful for this to return this information as well. Another option is to have a separate `getFQStack` or something, but I have a hard time imagining when you'd want to get a stack name that does not include org name as well.

@ellismg I will leave it to you to triage, but I believe we should fix this before we move `pulumi query` out from behind the debug flag.

**Comment by eropple** [comment]
There's a `name` field in `Pulumi.yaml` that seems like it should also be part of the qualified stack name. (This is particularly painful when trying to use the S3 backend, as right now all stacks in a single `pulumi login` are globally namespaced.)

**Comment by joeduffy** [comment]
This is related to https://github.com/pulumi/pulumi/issues/2622.

I'll admit that I was surprised where we landed after refactoring the engine to support the new project model. I assumed we would treat `org/stack` as the "stack name" everywhere, and then simply support a way of getting the simple name if that's all you needed. This pops up in more places than just `getStack`, like whenever we print the stack name in the CLI. But if we're going to stick with the opposite, we definitely do need a way of getting the fully qualified name.

I can imagine a myriad of options

1. Make `getStack` return the fully qualified `org/stack` name (breaking change)
2. Make `getStack` return an object with two properties, `.org` and `.stack` (breaking change)
3. Same as #2, but also add a few complementary functions (breaking change)
    - `getStackName` that returns `stack`
    - `getStackOrganization` that returns `org`
    - `getFullStackName` that returns `stack/org`
4. Add a new `getFullStackName` function that returns `stack/org`
5. Add a new `getOrganization` that returns `org`
6. _... eliding others ..._

I'll just mention passing by, as the one who made the mistake, that the `getStack` and `getProject` names are unfortunate because they sound like they return rich objects, but actually just return names.

**Comment by eropple** [comment]
The thing that I find awkward is that "org" is a Pulumi SaaS concept and doesn't map to a file-store backend (or similar, like S3).

I would expect, as a naive user, that a user would have SaaS logins that surfaced to them projects that they have visibility into, and projects would have stacks. S3 backends could then ignore the first step.

"Orgs" being a meaningful division (as opposed to a way to collect projects and apply ACLs to them, i.e. GitHub itself) is kinda odd.

**Comment by joeduffy** [comment]
I agree with the sentiment. I've always thought of org/stack as simply a path, which is actually the root cause for me expecting to see the fully qualified name everywhere -- a reasonable argument could be made that this should be an opaque string to the engine.

**Comment by keen99** [comment]
my vote would be a variant of #3 - but not breaking.  leave `getStack` as it is, add `getStackName` `getFullName` and `getOrgName` and `getAppName`   (we should definitely be able to query for the `Pulumi.yaml` `name:`

and `org` should be something that can be defined externally somewhere (right now that would probably be Pulumi.yaml or the stack.yaml unless the idea of a larger project scope config is added) if it's not from the backend (ie, non-saas backends)

it's important to not force the `org/app/stack` structure - a lot of use cases where you might want to programatically consume this can't support slashes (or other chars..), and requiring string replacement logic would suck worse than assembling ones own flavor with a few calls.  (ELB `name` is a good example)

**Comment by nesl247** [comment]
To chime in here, I would prefer not having `getStack()` be changed as all of our automation currently depends upon this as it is today. We have for example `Pulumi.development.yaml`, and thus when we run with `-s Org/development`, we depend upon `development` being returned in `getStack()` for automated updates.

**Comment by justinvp** [comment]
I'm inclined to only add `getFullStackName` for now. If we find there is an actual need to parse out the organization, we could consider adding an API that returns only that, but until then, I'd prefer to hold off since it doesn't make sense for other backends. (There's also a question about the naming, e.g. is it `getOrganizationName` or `getOwnerName`)?

One question about `getFullStackName`: For the Pulumi Service backend, does it return `org/stack` or `org/project/stack`?

**Comment by justinvp** [comment]
Let's consider this as part of any `query` work (where we'd likely make use of it).

**Comment by danfhernandez** [comment]
Any update or plan to implement this?  I think the easiest option would be to add another function that was mentioned by a few above: `getOrganization()` that returns `org`.  

@justinvp @joeduffy

**Comment by cnunciato** [comment]
There's a CLI use case for this as well -- e.g., I'd like to be able to set a value (e.g., in AWS Parameter Store) with the AWS CLI corresponding with the currently selected, fully-qualified stack name, such as `<org>/<project>/<name>`.

**Comment by vizv** [comment]
I'm currently using a hack :sweat_smile: :
```ts
import { execSync } from 'child_process'

const [organization] = execSync('pulumi login')
  .toString()
  .match(/(?<=Logged in to pulumi.com as )[^\s]+(?=.*)/)!
```

**Comment by slavafomin** [comment]
I believe there should be a separate method to get the organization name only.

For me this is useful when referencing other stacks:

```typescript
const stack = new pulumi.StackReference(`<org>/<project>/<stack>`);
```

I don't want to hardcode organization name in such calls, so being able to get the organization name would really help, e.g.:

```typescript
const orgName = pulumi.getOrgName();
const stageStack = new pulumi.StackReference(`${orgName}/foo/stage`);
```

**Comment by DusanDimitric** [comment]
@slavafomin I also see it as useful for referencing other stacks within your organization.

I am currently hardcoding the `ORGANIZATION` in a constant variable.

**Comment by Frassle** [comment]
The filestate backend doesn't support organisations. What should it return from getOrg? 

Little worried of we just return "" people won't check that and will make strings that look like "/proj/stack" which is invalid.

**Comment by joeduffy** [comment]
I wonder if it could be based on the underlying provider. For instance, the S3 one could use the bucket name as the org by default, and if you wanted to change it, we could support a `.pulumi` settings file in the bucket or something.

**Comment by phillipedwards** [comment]
`NotSupportedException` isn't great, but at least it's a quick, fail fast mechanism to stop people from using a `getOrganization` method when it doesn't apply (eg- filestate). This seems like a decent compromise to allow an org to be retrieved programmatically.

**Comment by AaronFriel** [comment]
I think we should change the API to return a constant value (TBD) for filestate backends to improve user experience:

Suppose I am an end user and I use a component provider, `pulumi new` a template, `git clone` a Pulumi program, or `npm install` a helper library. The author of the code I'm using used `getOrganization`, though I am just getting started and chose to use a filestate backend to try it out. That code, which could be authored by us, will crash their program and could leave resources in a pending state. That's nearly the worst possible getting started experience.

How do we mitigate that? Well, by making this function partial, we've imposed a cost on every author: you _must_ guard against failure by try-catch. We'll need to impose that cost on ourselves in blog posts, demos, sample projects, and component providers. It won't be a billion dollar mistake, but adding functions that throw into our API is _costly_.

I propose we return `"no-org"` or `"error-org-not-supported"` or some such.

**Comment by Frassle** [comment]
Need to add getOrg functions to Go and Java.

**Comment by Frassle** [comment]
I think every language now has functions to get the org, project, and stack names.
